File: app/src/main/java/no/nav/tilbakemeldingsmottak/rest/feilogmangler/FeilOgManglerRestController.java
package no.nav.tilbakemeldingsmottak.rest.feilogmangler;

import static no.nav.tilbakemeldingsmottak.metrics.MetricLabels.DOK_REQUEST;
import static no.nav.tilbakemeldingsmottak.metrics.MetricLabels.PROCESS_CODE;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.security.token.support.core.api.Protected;
import no.nav.tilbakemeldingsmottak.consumer.email.SendEmailException;
import no.nav.tilbakemeldingsmottak.metrics.Metrics;
import no.nav.tilbakemeldingsmottak.model.MeldFeilOgManglerRequest;
import no.nav.tilbakemeldingsmottak.model.MeldFeilOgManglerResponse;
import no.nav.tilbakemeldingsmottak.rest.feilogmangler.service.FeilOgManglerService;
import no.nav.tilbakemeldingsmottak.rest.feilogmangler.validation.MeldFeilOgManglerValidator;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;
import no.nav.tilbakemeldingsmottak.api.FeilOgManglerRestControllerApi;

import javax.transaction.Transactional;

@Protected
@RestController
@Slf4j
@RequiredArgsConstructor
public class FeilOgManglerRestController implements FeilOgManglerRestControllerApi {

    private final FeilOgManglerService feilOgManglerService;
    private final MeldFeilOgManglerValidator meldFeilOgManglerValidator;

    @Transactional
    @Override
    @Metrics(value = DOK_REQUEST, extraTags = {PROCESS_CODE, "meldFeilOgMangler"}, percentiles = {0.5, 0.95}, histogram = true)
    public ResponseEntity<MeldFeilOgManglerResponse> meldFeilOgMangler(@RequestBody MeldFeilOgManglerRequest request) throws SendEmailException {
            meldFeilOgManglerValidator.validateRequest(request);
            feilOgManglerService.meldFeilOgMangler(request);
            return ResponseEntity
                    .status(HttpStatus.OK)
                    .body(MeldFeilOgManglerResponse.builder().message("Feil/mangel meldt").build());
    }
}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/rest/feilogmangler/FeilOgManglerRestController.kt
package no.nav.tilbakemeldingsmottak.rest.feilogmangler

import no.nav.security.token.support.core.api.ProtectedWithClaims
import no.nav.tilbakemeldingsmottak.api.FeilOgManglerRestControllerApi
import no.nav.tilbakemeldingsmottak.config.Constants
import no.nav.tilbakemeldingsmottak.metrics.MetricLabels
import no.nav.tilbakemeldingsmottak.metrics.Metrics
import no.nav.tilbakemeldingsmottak.metrics.MetricsUtils
import no.nav.tilbakemeldingsmottak.model.MeldFeilOgManglerRequest
import no.nav.tilbakemeldingsmottak.model.MeldFeilOgManglerResponse
import no.nav.tilbakemeldingsmottak.rest.feilogmangler.service.FeilOgManglerService
import no.nav.tilbakemeldingsmottak.rest.feilogmangler.validation.MeldFeilOgManglerValidator
import no.nav.tilbakemeldingsmottak.util.OidcUtils
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@ProtectedWithClaims(issuer = "azuread")
@RestController
class FeilOgManglerRestController(
    private val feilOgManglerService: FeilOgManglerService,
    private val meldFeilOgManglerValidator: MeldFeilOgManglerValidator,
    private val metricsUtils: MetricsUtils,
    private val oidcUtils: OidcUtils
) : FeilOgManglerRestControllerApi {
    @Transactional
    @Metrics(
        value = MetricLabels.DOK_REQUEST,
        extraTags = [MetricLabels.PROCESS_CODE, "meldFeilOgMangler"],
        percentiles = [0.5, 0.95],
        histogram = true
    )
    override fun meldFeilOgMangler(@RequestBody meldFeilOgManglerRequest: MeldFeilOgManglerRequest): ResponseEntity<MeldFeilOgManglerResponse> {
        if (oidcUtils.getPidForIssuer(Constants.TOKENX_ISSUER) == null) {
            metricsUtils.incrementNotLoggedInRequestCounter(this.javaClass.name, "meldFeilOgMangler")
        }
        meldFeilOgManglerValidator.validateRequest(meldFeilOgManglerRequest)
        feilOgManglerService.meldFeilOgMangler(meldFeilOgManglerRequest)
        return ResponseEntity
            .status(HttpStatus.OK)
            .body(MeldFeilOgManglerResponse(message = "Feil/mangel meldt"))
    }
}
